### PR TITLE
Convert dashboard to progressive web app

### PR DESCRIPTION
**Pull request by lucasxsong**
## Proposed changes

Adding functionality to allow the existing dashboard to be used as a progressive web app (PWA). The current implementation uses the `vite-pwa` plugin and attaches a basic service worker during the build process. This provides the basic requirements needed to download the app as a PWA. You can test this by downloading the app on Chrome in the menubar as shown below.

Screenshot 2024-06-13 at 10 27 20 AM

The additions included should not impact any dashboard functionality. Additions that provide more offline support will be included in a later pull request. 

## Types of changes

What types of changes does this pull request introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (non-breaking change that does not add functionality but makes code cleaner or more efficient)
- [ ] Documentation Update
- [ ] Tests (new or updated tests)
- [ ] Style (changes to code styling)
- [ ] CI (continuous integration changes)
- [ ] Repository Maintenance
- [ ] Other (please describe below)

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/yeatmanlab/roar-dashboard/blob/main/.github/CONTRIBUTING.md).
- [x] The changes in this PR are as small as they can be. They represent one and only one fix or enhancement.
- [x] Linting checks pass with my changes.
- [x] Any existing unit tests pass with my changes.
- [x] Any existing end-to-end tests pass with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] If this PR fixes an existing issue, I have added a unit or end-to-end test that will detect if this issue reoccurs.
- [x] I have added JSDoc comments as appropriate.
- [x] I have added the necessary documentation to the [roar-docs repository](https://github.com/yeatmanlab/roar-docs).
- [x] I have shared this PR on the roar-pr-reviews channel (if I have access)
- [x] I have linked relevant issues (if any)

## Further comments

I've removed `vite-favicon-inject` in favor of just importing the favicons in our entry `index.html` file. I've also replaced the `basicSSL` plugin in favor of `mkcert` because I was getting an SSL error while installing the service worker with `basicSSL` that was remedied with using `mkcert`. I am unsure if this change has any other implications upstream, so please advise if you believe this update will affect any of our configurations.

### CODE REVIEW

**Review by kellyel — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -5,9 +5,19 @@
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
```
Comment (asengupta3) [inline index.html:5]: For a more optimized mobile view you can change this to:
```
<meta
      name="viewport"
      content="width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=1, viewport-fit=cover"
    />
```

File: index.html [hunk 2]
```diff
@@ -5,9 +5,19 @@
...
     <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin-allow-popups" />
     <title>ROAR</title>
+    <link rel="icon" href="public/favicon.ico" sizes="48x48" />
+    <link rel="icon" href="public/favicon.svg" sizes="any" type="image/svg+xml" />
+    <link rel="apple-touch-icon" href="/apple-touch-icon-180x180.png" />
```
Comment (asengupta3) [inline index.html:10]: Depending on which browsers you want to support here is a comprehensive list:
```
// Target ios browsers.
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
// Target safari on MacOS.
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
// The classic favicon displayed in tabs.
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
// Used by Android Chrome for the "Add to home screen" icon and settings.
<link rel="manifest" href="/site.webmanifest">
// Used for Safari pinned tabs.
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#193860">
// Target older browsers like IE 10 and lower.
<link rel="icon" href="/favicon.ico">
// Used by Chrome, Firefox OS, and opera to change the browser address bar.
<meta name="theme-color" content="#ccccc7">
// Used by windows 8, 8.1, and 10 for the start menu tiles.
<meta name="msapplication-TileColor" content="#00aba9">
```

File: index.html [hunk 3]
```diff
@@ -5,9 +5,19 @@
...
   </head>
   <!-- Google tag (gtag.js) -->
   <script async src="https://www.googletagmanager.com/gtag/js?id=G-YYE3YN0S99"></script>
+  <script>
+    if ('serviceWorker' in navigator) {
```
Comment (asengupta3) [inline index.html:15]: You may want to move this to a separate file and call it during the initialization of the app. This file in the future can do other things. Please take a look at an example of how this file usually looks here:
https://github.com/cra-template/pwa/blob/main/packages/cra-template-pwa-typescript/template/src/serviceWorkerRegistration.ts

File: vite.config.js [hunk 4]
```diff
@@ -1,19 +1,67 @@
...
+        ],
+        // Add more manifest options as needed
+      },
+      /* enable sw on development */
+      devOptions: {
+        enabled: true,
+        type: 'module',
+        /* other options */
+      },
+      registerType: 'autoUpdate',
```
Comment (asengupta3) [inline vite.config.js:62]: Autoupdate is easy but can cause inconvenience in production. When auto update triggers it does not save the current workflow and refreshes the page which can cause issues. I personally like the user to manually refresh it:
https://vite-pwa-org.netlify.app/guide/prompt-for-update.html
A lot of apps seem to follow this like Shortcut etc

**Comment by asengupta3** [comment]
@lucasxsong the above comments are suggestions from my experience with PWA. Not all need to be implemented but they are meant to elicit some thought on the subject.

**Review by asengupta3 — APPROVED** [review]
LGTM!

**Pull request merged**